wincent/liferay-portal
# LPS-110752 Create new modal utility

Hey @wincent ,

This is the initial implementation of modal utility, where the goal is to to keep API as similar as possible to AUI modals, but with implementation in ClayModal. See more [here](https://docs.google.com/document/d/10vUmbN3i1B02HaRh6RyYtLDSGhEgqmZvt0heGNiXIhI/edit#heading=h.zcgl7ezaxsmv).

In this PR, we are implementing the bare minimum needed to apply it to `segments-web`.

To test the code:
1. Go to Control Panel > Site > People > Segments > Add a segment
1. Add condition for a segment that first name is 'Test', save.
1. Click on 'View Members' button on top right. It should open the new modal, containing a list of users, in our case, the 'Test' user.

My uncertainties in this PR:
- We are rendering modal on `document.createElement('div')`, which is a detached node. `ClayModal` takes over and renders in `document.body`, I believe through its [`ClayPortal`](https://github.com/liferay/clay/blob/baf421c228a280eae38020efd5db52146546d8a5/packages/clay-modal/src/Modal.tsx#L78). I did not delve too deep to understand this solution. We want to render it it in `document.body`, so this behavior is fine, but it's unclear what to set in `render`. It needs to be set, it cannot be set to `document.body` as this eats siblings. We may add a container and remove it onClose, but it seems unnecessary, the default `ClayModal` behavior is perfectly fine. 
- On a related note, there is a part of API that enables setting of node where to render, `config.dialog.render`, but I'm not sure if there is a case where this is needed, we can likely just enforce that it always renders in `document.body`.
- Is there a more elegant way to apply `spritemap`? We have no context here provided in `<react:component>`, so I set it manually.
- Later down the line, there may be complications with implementations of `modalInstance.show()` and `modalInstance.hide()`. This is why my [initial sketch](https://github.com/markocikos/liferay-portal/commit/08e5ec17e51d601effee83d9c5ea3b377db13103) had modal as a class. We may return to that approach, we may drop the whole parent control of show / hide in usages, or we could find a way to extend the solution in this PR. For now, there is no need for it, I wanted to keep the most concise code, so I rewrote it. 

Please review the code.

Thank you!

/cc @jbalsas
